### DISCUSSION TITLE
SPFx 1.14 beta - can we get the MS Graph SDK updated?

### DISCUSSION THREAD

**Discussion by andrewconnell** [SharePoint/sp-dev-docs#7620] [Q&A] [answered]
The Microsoft Graph SDK included in the dependency tree is still using v1.1.0 that was specified when MS Graph support was added to SPFx in v1.6.0. That version is over 3 years old... the current version is v3... **can we get the dependency updated**?

Side benefit: developers shouldn't need to install the MS Graph JS SDK's type declarations as a separate package as they're included in the v3 SDK.

## TL;DR

The Microsoft Graph JS SDK [@microsoft/microsoft-graph-client](https://www.npmjs.com/package/@microsoft/microsoft-graph-client) is added to projects as a dependency by the **@microsoft/sp-http** package. That SPFx package explicitly references it ias `~1.1.0` which will only upgrade minor versions. As v1.1.0 is the last v1 major version published, developers are stuck using old tech.

```json
"dependencies": {
  "@microsoft/microsoft-graph-client": "~1.1.0",
}
```

Minimally, would be nice if the **@microsoft/sp-http** package's dependency specifies using the latest major version (v3, which is 5mo old at the time of this request):

```json
"dependencies": {
  "@microsoft/microsoft-graph-client": "^3.0.0"
}
```

Ideally, it gets all minor versions as well with each release:

```json
"dependencies": {
  "@microsoft/microsoft-graph-client": "~3.0.0"
}
```

**Comment by VesaJuvonen** [comment, marked as answer]
This is currently by design as the MS Graph SDK does not support running multiple versions of that package side-by-side on one page. This means that if you'd be using SPFx v1.0 and SPFx v1.14 in web parts on same page, there would be breaking issues for both web parts / components.

We would love to use newer version of the Microsoft Graph JS SDK in the SharePoint Framework, but are blocked until side-by-side is supported on the Microsoft Graph SDK side.

We are working though actively internally on this with the Microsoft Graph team, so you can expect this to change as soon as the side-by-side support is added in Graph side. This will not most likely happen by the SPFx 1.14 version, but let's see.

**Reply by andrewconnell** [reply]
Ah bummer… thanks for that detail. Wasn’t aware that’s what the hold up was.

**Reply by andrewconnell** [reply]
Do you know if there’s a tracking issue/link for this work item? Looked at their GH repo and can’t seem to find one.

**Reply by AJIXuMuK** [reply]
@andrewconnell - they don't have an issue on GH. But as Vesa mentioned we're working internally with Graph team to make it happen.

**Reply by sebastienlevert** [reply]
Thanks for sharing the details @VesaJuvonen! We are now identifying solutions with the SPFx team. @andrewconnell, how would you want to leverage the "right" version of the SDK in your code? Right now, to offer smaller package size, SPFx provides the Graph SDK and this is why you can't simply reference the v3 in your `package.json`. I'd love to get your input on this topic.

But to be clear, we 100% want to provide the latest and greatest SDK for SPFx developers.

**Reply by andrewconnell** [reply]
@sebastienlevert asked:
> how would you want to leverage the "right" version of the SDK in your code

I'd prefer to not have it included in my SPFx projects and let me include what I want. 

### TL;DR

There are ways to communicate "no support for conflicting versions on the same page".

As it stands now, even if I create a SPA in SPFx, I'm forced to use the v1.1.0 SDK. The same is true if I decide to build a-MS Teams tab with SPFx, or any other type of Teams component that can be built using SPFx, I'm blocked from using the latest SDK. 

**This is actually a compelling case to NOT use SPFx for MS Teams solutions that commonly need to get data from MS Graph.**

I'm curious though @sebastienlevert, does the MS Graph UI Toolkit include the SDK? If so, what happens if I use the latest MS Graph UI Toolkit within an SPFx solution? Does it blow up in my face?

**Reply by sebastienlevert** [reply]
Actually, MGT references the Graph SDK and we provide helpers to get tokens from SPFx. It does not blow up, it actually provides you the 2.x version of the SDK (We have an open PR for supporting v3, we have to check where we're at with this. I want it to happen ASAP). I blogged about it in April on how you can get a "fresher" version of the SDK via MGT (even without utilizing the components). This proves SDKs ca live side by side and we should look at more flexibility for the developer!

https://www.sebastienlevert.com/2021/04/18/latest-microsoft-graph-client-spfx/
